File: Day #466/solution.go
// Symmetric k-ary tree: recursively compare children of two subtrees in mirror order.
// Time: O(n), Space: O(h) recursion.
package main

import "fmt"

type Node struct {
	val      int
	children []*Node
}

func mirror(a, b *Node) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	if a.val != b.val {
		return false
	}
	if len(a.children) != len(b.children) {
		return false
	}
	n := len(a.children)
	for i := 0; i < n; i++ {
		if !mirror(a.children[i], b.children[n-1-i]) {
			return false
		}
	}
	return true
}

func isSymmetric(root *Node) bool {
	if root == nil {
		return true
	}
	return mirror(root, root)
}

func main() {
	root := &Node{val: 4}
	l3 := &Node{val: 3}
	m5 := &Node{val: 5}
	r3 := &Node{val: 3}
	root.children = []*Node{l3, m5, r3}
	l3.children = []*Node{{val: 9}}
	r3.children = []*Node{{val: 9}}
	if isSymmetric(root) {
		fmt.Println("True")
	} else {
		fmt.Println("False")
	}
}

File: Day #466/solution.rs
// Symmetric k-ary tree: recursively compare children of two subtrees in mirror order.
// Time: O(n), Space: O(h) recursion.
struct Node {
    val: i32,
    children: Vec<Node>,
}

fn mirror(a: &Node, b: &Node) -> bool {
    if a.val != b.val {
        return false;
    }
    if a.children.len() != b.children.len() {
        return false;
    }
    let n = a.children.len();
    for i in 0..n {
        if !mirror(&a.children[i], &b.children[n - 1 - i]) {
            return false;
        }
    }
    true
}

fn is_symmetric(root: &Node) -> bool {
    mirror(root, root)
}

fn main() {
    let root = Node {
        val: 4,
        children: vec![
            Node { val: 3, children: vec![Node { val: 9, children: vec![] }] },
            Node { val: 5, children: vec![] },
            Node { val: 3, children: vec![Node { val: 9, children: vec![] }] },
        ],
    };
    println!("{}", if is_symmetric(&root) { "True" } else { "False" });
}
